File: src/dataManager.js
/**
 * FM24 Scout - Data Manager
 * Handles loading and managing Football Manager roles and player data
 */

import rolesData from '../data/roles.json';

// Cache for loaded data
let rolesCache = null;
let playersCache = [];

/**
 * Initialize and load all role data
 * @returns {Array} Array of role objects
 */
export function loadRoles() {
  if (!rolesCache) {
    rolesCache = rolesData;
    console.log(`Loaded ${rolesCache.length} tactical roles`);
  }
  return rolesCache;
}

/**
 * Get a specific role by its code
 * @param {string} code - Role code (e.g., 'afa' for Advanced Forward Attack)
 * @returns {Object|null} Role object or null if not found
 */
export function getRoleByCode(code) {
  const roles = loadRoles();
  return roles.find(role => role.RoleCode === code) || null;
}

/**
 * Get all roles for a specific duty
 * @param {string} duty - Duty type ('Attack', 'Support', 'Defend')
 * @returns {Array} Array of matching roles
 */
export function getRolesByDuty(duty) {
  const roles = loadRoles();
  return roles.filter(role => role.Role.includes(duty));
}

/**
 * Get all available role names
 * @returns {Array} Array of role names
 */
export function getAllRoleNames() {
  const roles = loadRoles();
  return roles.map(role => ({
    name: role.Role,
    code: role.RoleCode
  }));
}

/**
 * Get all unique attributes from roles
 * @returns {Array} Array of attribute names
 */
export function getAllAttributes() {
  const roles = loadRoles();
  if (roles.length === 0) return [];
  
  return Object.keys(roles[0])
    .filter(key => key !== 'Role' && key !== 'RoleCode')
    .sort();
}

/**
 * Parse HTML file content to extract player data
 * @param {string} htmlContent - HTML file content
 * @returns {Object} Parsed player data
 */
export function parseHTMLFile(htmlContent) {
  const parser = new DOMParser();
  const doc = parser.parseFromString(htmlContent, 'text/html');
  const table = doc.querySelector('table');
  
  if (!table) {
    throw new Error('No table found in HTML file');
  }
  
  // Extract headers
  const headers = Array.from(table.querySelectorAll('thead th'))
    .map(th => th.textContent.trim());
  
  if (headers.length === 0) {
    // Try first row as headers
    const firstRow = table.querySelector('tr');
    if (firstRow) {
      headers.push(...Array.from(firstRow.querySelectorAll('th, td'))
        .map(cell => cell.textContent.trim()));
    }
  }
  
  // Extract player data
  const rows = Array.from(table.querySelectorAll('tbody tr'))
    .map(tr => {
      const cells = Array.from(tr.querySelectorAll('td'));
      const player = {};
      
      headers.forEach((header, index) => {
        const value = cells[index]?.textContent.trim() || '-';
        player[header] = parseAttributeValue(value);
      });
      
      // Add calculated attributes
      player.Speed = calculateSpeed(player);
      player.WorkRate = calculateWorkRate(player);
      player.SetPieces = calculateSetPieces(player);
      
      return player;
    });
  
  // Validate we have player data
  const validRows = rows.filter(player => 
    player.Name && player.Name !== '-'
  );
  
  if (validRows.length === 0) {
    throw new Error('No valid player data found in table');
  }
  
  if (validRows.length > 20000) {
    throw new Error(`File too large: ${validRows.length} players found (maximum 20,000)`);
  }
  
  return {
    headers,
    players: validRows,
    count: validRows.length
  };
}

/**
 * Parse CSV file content to extract player data
 * @param {string} csvContent - CSV file content
 * @returns {Object} Parsed player data
 */
export function parseCSVFile(csvContent) {
  const lines = csvContent.split(/\r?\n/);
  const headers = lines[0].split(',').map(h => h.trim().replace(/^"|"$/g, ''));
  
  const players = lines.slice(1)
    .filter(line => line.trim())
    .map(line => {
      // Handle CSV with potential commas in quoted values
      const values = parseCSVLine(line);
      const player = {};
      
      headers.forEach((header, index) => {
        const value = values[index]?.trim() || '-';
        player[header] = parseAttributeValue(value);
      });
      
      // Add calculated attributes
      player.Speed = calculateSpeed(player);
      player.WorkRate = calculateWorkRate(player);
      player.SetPieces = calculateSetPieces(player);
      
      return player;
    })
    .filter(player => player.Name && player.Name !== '-');
  
  if (players.length === 0) {
    throw new Error('No valid player data found in CSV');
  }
  
  if (players.length > 20000) {
    throw new Error(`File too large: ${players.length} players found (maximum 20,000)`);
  }
  
  return {
    headers,
    players,
    count: players.length
  };
}

/**
 * Parse a single CSV line handling quoted values
 * @param {string} line - CSV line
 * @returns {Array} Array of values
 */
function parseCSVLine(line) {
  const values = [];
  let current = '';
  let inQuotes = false;
  
  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    
    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      values.push(current);
      current = '';
    } else {
      current += char;
    }
  }
  
  values.push(current);
  return values.map(v => v.replace(/^"|"$/g, ''));
}

/**
 * Parse attribute value handling ranges and special cases
 * @param {string} value - Raw attribute value
 * @returns {number|string} Parsed value
 */
function parseAttributeValue(value) {
  // Handle missing values
  if (value === '-' || value === '' || value === null || value === undefined) {
    return 0;
  }
  
  // Handle range values (e.g., "14-16")
  if (typeof value === 'string' && value.includes('-') && !value.startsWith('-')) {
    const parts = value.split('-').map(p => parseFloat(p.trim()));
    if (parts.length === 2 && !isNaN(parts[0]) && !isNaN(parts[1])) {
      return (parts[0] + parts[1]) / 2;
    }
  }
  
  // Try to parse as number
  const num = parseFloat(value);
  if (!isNaN(num)) {
    return num;
  }
  
  // Return as string if not a number
  return value;
}

/**
 * Calculate Speed from Pace and Acceleration
 * @param {Object} player - Player object
 * @returns {number} Calculated speed
 */
function calculateSpeed(player) {
  const pace = parseFloat(player.Pac) || 0;
  const acceleration = parseFloat(player.Acc) || 0;
  return Math.round((pace + acceleration) / 2 * 10) / 10;
}

/**
 * Calculate Work Rate from Work Rate and Stamina
 * @param {Object} player - Player object
 * @returns {number} Calculated work rate
 */
function calculateWorkRate(player) {
  const workRate = parseFloat(player.Wor) || 0;
  const stamina = parseFloat(player.Sta) || 0;
  return Math.round((workRate + stamina) / 2 * 10) / 10;
}

/**
 * Calculate Set Pieces average
 * @param {Object} player - Player object
 * @returns {number} Calculated set pieces value
 */
function calculateSetPieces(player) {
  const corners = parseFloat(player.Cor) || 0;
  const freeKicks = parseFloat(player.Fre) || 0;
  const penalties = parseFloat(player.Pen) || 0;
  const throwIns = parseFloat(player.Thr) || 0;
  
  const values = [corners, freeKicks, penalties, throwIns].filter(v => v > 0);
  
  if (values.length === 0) return 0;
  
  const sum = values.reduce((a, b) => a + b, 0);
  return Math.round(sum / values.length * 10) / 10;
}

/**
 * Calculate player scores for all roles
 * @param {Array} players - Array of player objects
 * @param {Array} selectedRoles - Array of role codes to calculate
 * @returns {Array} Players with calculated role scores
 */
export function calculatePlayerScores(players, selectedRoles = null) {
  const roles = loadRoles();
  const rolesToCalculate = selectedRoles 
    ? roles.filter(r => selectedRoles.includes(r.RoleCode))
    : roles;
  
  return players.map(player => {
    const roleScores = {};
    
    rolesToCalculate.forEach(role => {
      let totalScore = 0;
      let weightSum = 0;
      
      // Calculate weighted score for each attribute
      Object.keys(role).forEach(attr => {
        if (attr !== 'Role' && attr !== 'RoleCode') {
          const weight = role[attr];
          const playerValue = parseFloat(player[attr]) || 0;
          
          if (weight > 0) {
            // Normalize player value (assuming max attribute value is 20)
            const normalizedValue = Math.min(playerValue / 20, 1);
            totalScore += normalizedValue * weight;
            weightSum += weight;
          }
        }
      });
      
      // Calculate percentage score
      roleScores[role.RoleCode] = weightSum > 0 
        ? Math.round((totalScore / weightSum) * 100 * 100) / 100
        : 0;
    });
    
    // Find best roles
    const sortedRoles = Object.entries(roleScores)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5);
    
    return {
      ...player,
      roleScores,
      bestRoles: sortedRoles.map(([code, score]) => ({
        code,
        name: roles.find(r => r.RoleCode === code)?.Role,
        score
      }))
    };
  });
}

/**
 * Store players in cache
 * @param {Array} players - Array of player objects
 */
export function cachePlayers(players) {
  playersCache = players;
  console.log(`Cached ${players.length} players`);
}

/**
 * Get cached players
 * @returns {Array} Cached players array
 */
export function getCachedPlayers() {
  return playersCache;
}

/**
 * Clear all cached data
 */
export function clearCache() {
  playersCache = [];
  console.log('Player cache cleared');
}

// Export for use in other modules
export default {
  loadRoles,
  getRoleByCode,
  getRolesByDuty,
  getAllRoleNames,
  getAllAttributes,
  parseHTMLFile,
  parseCSVFile,
  calculatePlayerScores,
  cachePlayers,
  getCachedPlayers,
  clearCache
};

File: src-tauri/src/data_manager.rs
// FM24 Scout - Rust Data Manager
// Handles role data and player file processing in the Tauri backend

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use tauri::command;

// Role structure matching the JSON data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Role {
    #[serde(rename = "Role")]
    pub role: String,
    #[serde(rename = "RoleCode")]
    pub role_code: String,
    #[serde(rename = "1v1")]
    pub one_v_one: u8,
    #[serde(rename = "Acc")]
    pub acceleration: u8,
    #[serde(rename = "Aer")]
    pub aerial_reach: u8,
    #[serde(rename = "Agg")]
    pub aggression: u8,
    #[serde(rename = "Agi")]
    pub agility: u8,
    #[serde(rename = "Ant")]
    pub anticipation: u8,
    #[serde(rename = "Bal")]
    pub balance: u8,
    #[serde(rename = "Bra")]
    pub bravery: u8,
    #[serde(rename = "Cmd")]
    pub command_of_area: u8,
    #[serde(rename = "Cnt")]
    pub concentration: u8,
    #[serde(rename = "Cmp")]
    pub composure: u8,
    #[serde(rename = "Cro")]
    pub crossing: u8,
    #[serde(rename = "Dec")]
    pub decisions: u8,
    #[serde(rename = "Det")]
    pub determination: u8,
    #[serde(rename = "Dri")]
    pub dribbling: u8,
    #[serde(rename = "Fin")]
    pub finishing: u8,
    #[serde(rename = "Fir")]
    pub first_touch: u8,
    #[serde(rename = "Fla")]
    pub flair: u8,
    #[serde(rename = "Han")]
    pub handling: u8,
    #[serde(rename = "Hea")]
    pub heading: u8,
    #[serde(rename = "Jum")]
    pub jumping_reach: u8,
    #[serde(rename = "Kic")]
    pub kicking: u8,
    #[serde(rename = "Ldr")]
    pub leadership: u8,
    #[serde(rename = "Lon")]
    pub long_shots: u8,
    #[serde(rename = "Mar")]
    pub marking: u8,
    #[serde(rename = "OtB")]
    pub off_the_ball: u8,
    #[serde(rename = "Pac")]
    pub pace: u8,
    #[serde(rename = "Pas")]
    pub passing: u8,
    #[serde(rename = "Pos")]
    pub positioning: u8,
    #[serde(rename = "Ref")]
    pub reflexes: u8,
    #[serde(rename = "Sta")]
    pub stamina: u8,
    #[serde(rename = "Str")]
    pub strength: u8,
    #[serde(rename = "Tck")]
    pub tackling: u8,
    #[serde(rename = "Tea")]
    pub teamwork: u8,
    #[serde(rename = "Tec")]
    pub technique: u8,
    #[serde(rename = "Thr")]
    pub throwing: u8,
    #[serde(rename = "TRO")]
    pub tendency_to_rush_out: u8,
    #[serde(rename = "Vis")]
    pub vision: u8,
    #[serde(rename = "Wor")]
    pub work_rate: u8,
    #[serde(rename = "Cor")]
    pub corners: u8,
}

// Player data structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Player {
    pub name: String,
    pub nationality: String,
    pub club: String,
    pub position: String,
    pub attributes: HashMap<String, f32>,
    pub calculated_attributes: CalculatedAttributes,
    pub role_scores: HashMap<String, f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CalculatedAttributes {
    pub speed: f32,
    pub work_rate: f32,
    pub set_pieces: f32,
}

// Import result structure
#[derive(Debug, Serialize, Deserialize)]
pub struct ImportResult {
    pub success: bool,
    pub player_count: usize,
    pub players: Vec<Player>,
    pub error: Option<String>,
}

// Embed the roles.json file at compile time
const ROLES_JSON: &str = include_str!("../data/roles.json");

// Static roles data loaded once
lazy_static::lazy_static! {
    static ref ROLES: Vec<Role> = {
        serde_json::from_str(ROLES_JSON)
            .expect("Failed to parse embedded roles.json")
    };
}

// Get all roles
#[command]
pub fn get_roles() -> Vec<Role> {
    ROLES.clone()
}

// Get role by code
#[command]
pub fn get_role_by_code(code: String) -> Option<Role> {
    ROLES.iter().find(|r| r.role_code == code).cloned()
}

// Get roles by duty
#[command]
pub fn get_roles_by_duty(duty: String) -> Vec<Role> {
    ROLES
        .iter()
        .filter(|r| r.role.contains(&duty))
        .cloned()
        .collect()
}

// Import player file (HTML or CSV)
#[command]
pub async fn import_player_file(file_path: String) -> Result<ImportResult, String> {
    let content = fs::read_to_string(&file_path)
        .map_err(|e| format!("Failed to read file: {}", e))?;
    
    let players = if file_path.ends_with(".html") || file_path.ends_with(".htm") {
        parse_html_file(&content)?
    } else if file_path.ends_with(".csv") {
        parse_csv_file(&content)?
    } else {
        return Err("Unsupported file format. Please use HTML or CSV files.".to_string());
    };
    
    // Check player count limit
    if players.len() > 20000 {
        return Err(format!(
            "File too large: {} players found (maximum 20,000)",
            players.len()
        ));
    }
    
    Ok(ImportResult {
        success: true,
        player_count: players.len(),
        players,
        error: None,
    })
}

// Parse HTML file
fn parse_html_file(content: &str) -> Result<Vec<Player>, String> {
    use scraper::{Html, Selector};
    
    let document = Html::parse_document(content);
    let table_selector = Selector::parse("table").unwrap();
    let row_selector = Selector::parse("tr").unwrap();
    let cell_selector = Selector::parse("td, th").unwrap();
    
    let table = document
        .select(&table_selector)
        .next()
        .ok_or("No table found in HTML file")?;
    
    let mut rows = table.select(&row_selector);
    
    // Get headers from first row
    let header_row = rows.next().ok_or("No header row found")?;
    let headers: Vec<String> = header_row
        .select(&cell_selector)
        .map(|cell| cell.text().collect::<String>().trim().to_string())
        .collect();
    
    // Parse player rows
    let mut players = Vec::new();
    
    for row in rows {
        let cells: Vec<String> = row
            .select(&cell_selector)
            .map(|cell| cell.text().collect::<String>().trim().to_string())
            .collect();
        
        if cells.len() != headers.len() {
            continue;
        }
        
        let mut attributes = HashMap::new();
        let mut name = String::new();
        let mut nationality = String::new();
        let mut club = String::new();
        let mut position = String::new();
        
        for (i, header) in headers.iter().enumerate() {
            let value = &cells[i];
            
            match header.as_str() {
                "Name" => name = value.clone(),
                "Nationality" => nationality = value.clone(),
                "Club" => club = value.clone(),
                "Position" => position = value.clone(),
                _ => {
                    let parsed_value = parse_attribute_value(value);
                    attributes.insert(header.clone(), parsed_value);
                }
            }
        }
        
        if name.is_empty() || name == "-" {
            continue;
        }
        
        let calculated_attributes = calculate_derived_attributes(&attributes);
        let role_scores = calculate_role_scores(&attributes);
        
        players.push(Player {
            name,
            nationality,
            club,
            position,
            attributes,
            calculated_attributes,
            role_scores,
        });
    }
    
    if players.is_empty() {
        return Err("No valid player data found in HTML file".to_string());
    }
    
    Ok(players)
}

// Parse CSV file
fn parse_csv_file(content: &str) -> Result<Vec<Player>, String> {
    let mut reader = csv::Reader::from_reader(content.as_bytes());
    let headers = reader
        .headers()
        .map_err(|e| format!("Failed to read CSV headers: {}", e))?
        .clone();
    
    let mut players = Vec::new();
    
    for result in reader.records() {
        let record = result.map_err(|e| format!("Failed to read CSV record: {}", e))?;
        
        let mut attributes = HashMap::new();
        let mut name = String::new();
        let mut nationality = String::new();
        let mut club = String::new();
        let mut position = String::new();
        
        for (i, field) in record.iter().enumerate() {
            if let Some(header) = headers.get(i) {
                match header {
                    "Name" => name = field.to_string(),
                    "Nationality" => nationality = field.to_string(),
                    "Club" => club = field.to_string(),
                    "Position" => position = field.to_string(),
                    _ => {
                        let parsed_value = parse_attribute_value(field);
                        attributes.insert(header.to_string(), parsed_value);
                    }
                }
            }
        }
        
        if name.is_empty() || name == "-" {
            continue;
        }
        
        let calculated_attributes = calculate_derived_attributes(&attributes);
        let role_scores = calculate_role_scores(&attributes);
        
        players.push(Player {
            name,
            nationality,
            club,
            position,
            attributes,
            calculated_attributes,
            role_scores,
        });
    }
    
    if players.is_empty() {
        return Err("No valid player data found in CSV file".to_string());
    }
    
    Ok(players)
}

// Parse attribute value (handle ranges and missing values)
fn parse_attribute_value(value: &str) -> f32 {
    if value == "-" || value.is_empty() {
        return 0.0;
    }
    
    // Handle range values like "14-16"
    if value.contains('-') && !value.starts_with('-') {
        let parts: Vec<&str> = value.split('-').collect();
        if parts.len() == 2 {
            if let (Ok(min), Ok(max)) = (parts[0].parse::<f32>(), parts[1].parse::<f32>()) {
                return (min + max) / 2.0;
            }
        }
    }
    
    value.parse::<f32>().unwrap_or(0.0)
}

// Calculate derived attributes
fn calculate_derived_attributes(attributes: &HashMap<String, f32>) -> CalculatedAttributes {
    let pace = attributes.get("Pac").unwrap_or(&0.0);
    let acceleration = attributes.get("Acc").unwrap_or(&0.0);
    let speed = (pace + acceleration) / 2.0;
    
    let work_rate_attr = attributes.get("Wor").unwrap_or(&0.0);
    let stamina = attributes.get("Sta").unwrap_or(&0.0);
    let work_rate = (work_rate_attr + stamina) / 2.0;
    
    let corners = attributes.get("Cor").unwrap_or(&0.0);
    let free_kicks = attributes.get("Fre").unwrap_or(&0.0);
    let penalties = attributes.get("Pen").unwrap_or(&0.0);
    let throw_ins = attributes.get("Thr").unwrap_or(&0.0);
    
    let set_pieces_values: Vec<f32> = vec![*corners, *free_kicks, *penalties, *throw_ins]
        .into_iter()
        .filter(|&v| v > 0.0)
        .collect();
    
    let set_pieces = if set_pieces_values.is_empty() {
        0.0
    } else {
        set_pieces_values.iter().sum::<f32>() / set_pieces_values.len() as f32
    };
    
    CalculatedAttributes {
        speed,
        work_rate,
        set_pieces,
    }
}

// Calculate role scores for a player
fn calculate_role_scores(attributes: &HashMap<String, f32>) -> HashMap<String, f32> {
    let mut scores = HashMap::new();
    
    for role in ROLES.iter() {
        let mut total_score = 0.0;
        let mut weight_sum = 0.0;
        
        // Map role attributes to player attributes
        let role_weights = vec![
            ("1v1", role.one_v_one),
            ("Acc", role.acceleration),
            ("Aer", role.aerial_reach),
            ("Agg", role.aggression),
            ("Agi", role.agility),
            ("Ant", role.anticipation),
            ("Bal", role.balance),
            ("Bra", role.bravery),
            ("Cmd", role.command_of_area),
            ("Cnt", role.concentration),
            ("Cmp", role.composure),
            ("Cro", role.crossing),
            ("Dec", role.decisions),
            ("Det", role.determination),
            ("Dri", role.dribbling),
            ("Fin", role.finishing),
            ("Fir", role.first_touch),
            ("Fla", role.flair),
            ("Han", role.handling),
            ("Hea", role.heading),
            ("Jum", role.jumping_reach),
            ("Kic", role.kicking),
            ("Ldr", role.leadership),
            ("Lon", role.long_shots),
            ("Mar", role.marking),
            ("OtB", role.off_the_ball),
            ("Pac", role.pace),
            ("Pas", role.passing),
            ("Pos", role.positioning),
            ("Ref", role.reflexes),
            ("Sta", role.stamina),
            ("Str", role.strength),
            ("Tck", role.tackling),
            ("Tea", role.teamwork),
            ("Tec", role.technique),
            ("Thr", role.throwing),
            ("TRO", role.tendency_to_rush_out),
            ("Vis", role.vision),
            ("Wor", role.work_rate),
            ("Cor", role.corners),
        ];
        
        for (attr_name, weight) in role_weights {
            if weight > 0 {
                let player_value = attributes.get(attr_name).unwrap_or(&0.0);
                // Normalize to 0-1 range (assuming max attribute is 20)
                let normalized_value = (player_value / 20.0).min(1.0);
                total_score += normalized_value * weight as f32;
                weight_sum += weight as f32;
            }
        }
        
        let score = if weight_sum > 0.0 {
            (total_score / weight_sum) * 100.0
        } else {
            0.0
        };
        
        scores.insert(role.role_code.clone(), score);
    }
    
    scores
}

// Calculate scores for selected roles only
#[command]
pub fn calculate_player_scores(
    players: Vec<HashMap<String, serde_json::Value>>,
    selected_roles: Vec<String>,
) -> Result<Vec<HashMap<String, serde_json::Value>>, String> {
    let mut results = Vec::new();
    
    for player_data in players {
        let mut attributes = HashMap::new();
        
        // Extract attributes from player data
        for (key, value) in &player_data {
            if let Some(num) = value.as_f64() {
                attributes.insert(key.clone(), num as f32);
            } else if let Some(s) = value.as_str() {
                attributes.insert(key.clone(), parse_attribute_value(s));
            }
        }
        
        // Calculate scores for selected roles
        let mut role_scores = HashMap::new();
        for role_code in &selected_roles {
            if let Some(role) = ROLES.iter().find(|r| r.role_code == *role_code) {
                let score = calculate_single_role_score(&attributes, role);
                role_scores.insert(role_code.clone(), score);
            }
        }
        
        // Add scores to player data
        let mut result = player_data.clone();
        result.insert(
            "roleScores".to_string(),
            serde_json::to_value(role_scores).unwrap(),
        );
        
        results.push(result);
    }
    
    Ok(results)
}

// Calculate score for a single role
fn calculate_single_role_score(attributes: &HashMap<String, f32>, role: &Role) -> f32 {
    let mut total_score = 0.0;
    let mut weight_sum = 0.0;
    
    let role_weights = vec![
        ("1v1", role.one_v_one),
        ("Acc", role.acceleration),
        ("Aer", role.aerial_reach),
        // ... (same as above)
    ];
    
    for (attr_name, weight) in role_weights {
        if weight > 0 {
            let player_value = attributes.get(attr_name).unwrap_or(&0.0);
            let normalized_value = (player_value / 20.0).min(1.0);
            total_score += normalized_value * weight as f32;
            weight_sum += weight as f32;
        }
    }
    
    if weight_sum > 0.0 {
        (total_score / weight_sum) * 100.0
    } else {
        0.0
    }
}